dynamic_programming/knapsack.c

```diff
@@ -0,0 +1,184 @@
+ * Given a set of items, each with a weight and a value, determine
+ * which items to include in the collection
+ * so that the total weight is less than or equal to a given limit 
+ * and the total value is as large as possible.
+ * @author [aa001R](https://github.com/aa001R)
+ */
+
+#include <assert.h>  /// for assert
+#include <stdio.h>
+
+/**
+ * @brief max() macro function
+ */
+#define MAX(a, b) (((a) > (b)) ? (a) : (b))
+
+/**
+ * @brief Maximum capacity of the knapsack
+ */
+#define MAXCAPACITY 10
+
+/**
+ * @brief Item structure
+ */
+struct Item
+{
+    int weight;  // weight
+    int value;   // value
+};
+
+/**
+ * @brief Print input items function
+ * @param items an array of Item structure
+ * @param itemCount the number of items(Item structure)
+ * @returns void
+ */
+void print_input_struct_Items(struct Item items[], int itemCount);
+/**
+ * @brief Print the combination of items in the knapsack
+ * @param maxWeight the capacity of the knapsack
+ * @param items an array of Item structure
+ * @param itemCount the number of items(Item structure)
+ * @param K the 2D array storing the optimal solution
+ * @returns void
+ */
+void print_knapsack_item(int maxWeight, struct Item items[], int itemCount,
+                         int K[][MAXCAPACITY + 1]);
+
+/**
+ * @brief 0-1 Knapsack algorithm
+ * @param maxWeight the capacity of the knapsack
+ * @param items an array of Item structure
+ * @param itemCount the number of items(Item structure)
+ * @returns the maximum value that can be obtained
+ */
+int knapsack(int maxWeight, struct Item items[], int itemCount)
+{
+    // Print input items
+    print_input_struct_Items(items, itemCount);
+
+    // 2D array to store the optimal solution of subproblems
+    int K[itemCount + 1][maxWeight + 1];
+
+    // Initialize the first row and column to 0
+    for (int i = 0; i <= itemCount; i++) K[i][0] = 0;
+    for (int j = 0; j <= maxWeight; j++) K[0][j] = 0;
+
+    // Calculate the optimal solution by selecting items one by one
+    for (int i = 1; i <= itemCount; i++)
+    {
+        for (int w = 1; w <= maxWeight; w++)
+        {
+            if (items[i - 1].weight <= w)
+            {  // If the i-th item can be included in the knapsack
+                int valueWithPick =
+                    items[i - 1].value +
+                    K[i - 1][w - items[i - 1].weight];  // Value if the i-th
+                                                        // item is selected
+                int valueWithoutPick =
+                    K[i - 1][w];  // Value if the i-th item is not selected
+                K[i][w] = MAX(valueWithPick,
+                              valueWithoutPick);  // Select the maximum value
+            }
+            else
+            {  // If the i-th item cannot be included in the knapsack
+                K[i][w] = K[i - 1][w];
+            }
+        }
+    }
+
+    // Print the result of the Knapsack algorithm
+    printf("Knapsack execution result:\n");
```

Comment:
probably it would be better to separate calculating and printing in to defferent methods?

Reply:
Thank you for the suggestion:) 

I have added a new method called print_knapsack_result() specifically for the printing operation !